lib/segment/src/data_types/order_by.rs

```diff
@@ -81,3 +85,60 @@ impl OrderBy {
             })
     }
 }
+
+#[derive(Eq)]
+pub enum OrderingValue {
+    Float(OrderedFloat<FloatPayloadType>),
+    Int(i64),
+}
+
+impl From<OrderingValue> for serde_json::Value {
+    fn from(value: OrderingValue) -> Self {
+        match value {
+            OrderingValue::Float(value) => serde_json::Number::from_f64(value.into_inner())
+                .map(serde_json::Value::Number)
+                .unwrap_or(serde_json::Value::Null),
+            OrderingValue::Int(value) => serde_json::Value::Number(serde_json::Number::from(value)),
+        }
+    }
+}
+
+impl From<FloatPayloadType> for OrderingValue {
+    fn from(value: FloatPayloadType) -> Self {
+        OrderingValue::Float(OrderedFloat(value))
+    }
+}
+
+impl From<i64> for OrderingValue {
+    fn from(value: i64) -> Self {
+        OrderingValue::Int(value)
+    }
+}
+
+impl PartialEq for OrderingValue {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (OrderingValue::Float(a), OrderingValue::Float(b)) => a == b,
+            (OrderingValue::Int(a), OrderingValue::Int(b)) => a == b,
+            (OrderingValue::Float(a), OrderingValue::Int(b)) => a == &OrderedFloat(*b as f64),
```

Comment:
Lossy type conversion violates the transitive condition required by `PartialEq` and `PartialCmp`. (see [docs](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html))

```
let a = OrderingValue::from(9007199254740992);
let b = OrderingValue::from(9007199254740992.0);
let c = OrderingValue::from(9007199254740993);
assert!(a == b); // true
assert!(b == c); // true
assert!(a == c); // false
```

I think to solve it we could provide a function `fn(i64, f64) -> Ordering` that carefully handles all possible cases; and reuse this function in `PartialEq` and `Ord` implementations. Maybe it's already implemented in some crate.

---

Not related to your PR, but a fun trivia. A recent vulnerability found in glibc: if you pass a non-transitive comparison function to `qsort`, you'll get out-of-bounds read and writes. https://www.openwall.com/lists/oss-security/2024/01/30/7
It would be not that severe in our case, but I think it still could break sorting or other function's logic.

Reply:
Because of f64 rounding. It could be simplified to this:
```rust
assert!(9007199254740992.0 == 9007199254740993.0); // true
```
`2**53 + 1.0` could not be represented by f64, so it's getting rounded to `2**53`.

Reply:
looks like the violation of transitive condition is only a consequence of comparison which works incorrectly then

Reply:
It works correctly since `9007199254740992.0f64` has the same representation as `9007199254740993.0f64`. It's the same value bitwise. (`println!("{}", 9007199254740993.0);` prints the `…92` not `…93`)

The problem arise because of the cast (`as f64`) being used. The rounding happens in it.

Reply:
maybe this https://crates.io/crates/num-cmp ?

Reply:
From a quick glance and testing this crate does the right thing.

Reply:
I did update the code to follow the `NaN` handling, in the same way as `OrderingFloat` does.